pkg/gktest/read_suites.go

```diff
@@ -1,18 +1,175 @@
 package gktest
 
-import "io/fs"
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var (
+	// ErrNoFileSystem means a method which expects a filesystem got nil
+	// instead. This is likely a bug in the code.
+	ErrNoFileSystem = errors.New("no filesystem")
+	// ErrNoTarget indicates that the user did not specify a target directory or
+	// file.
+	ErrNoTarget = errors.New("target not specified")
+	// ErrUnsupportedExtension indicates that a user attempted to run tests in
+	// a file type which is not supported.
+	ErrUnsupportedExtension = errors.New("unsupported extension")
+	// ErrInvalidYAML indicates that a .yaml/.yml file was not parseable.
+	ErrInvalidYAML = errors.New("invalid yaml")
+	// ErrNotADirectory indicates that a user is mistakenly attempting to
+	// perform a directory-only action on a file (for example, recursively
+	// traversing it)
+	ErrNotADirectory = errors.New("not a directory")
+)
+
+const (
+	// Group is the API Group for Test YAML objects.
+	Group = "test.gatekeeper.sh"
+	// Kind is the Kind for Suite YAML objects.
+	Kind = "Suite"
+)
 
 // ReadSuites returns the set of test Suites selected by path.
 //
 // 1) If path is a path to a Suite, parses and returns the Suite.
-// 2) If the path is a directory, returns the Suites defined in that directory
-//    (not recursively).
-// 3) If the path is a directory followed by "...", returns all Suites in that
+// 2) If the path is a directory and recursive is false, returns only the Suites
+//    defined in that directory.
+// 3) If the path is a directory and recursive is true returns all Suites in that
 //      directory and its subdirectories.
 //
 // Returns an error if:
 // - path is a file that does not define a Suite
 // - any matched files containing Suites are not parseable
-func ReadSuites(f fs.FS, path string) ([]Suite, error) {
-	return nil, nil
+func ReadSuites(f fs.FS, target string, recursive bool) ([]Suite, error) {
+	if f == nil {
+		return nil, ErrNoFileSystem
+	}
+	if target == "" {
+		return nil, ErrNoTarget
+	}
+
+	stat, err := fs.Stat(f, target)
+	if err != nil {
+		return nil, err
+	}
+
+	files := fileList{}
+	switch {
+	case !stat.IsDir() && !recursive:
+		// target is a file.
+		err = files.addFile(target)
+
+	case !stat.IsDir() && recursive:
+		// target is a file, but user specified it should be traversed recursively.
```

Comment:
nit: "the user specified" to differentiate between "user-specified", for this and subsequent instances

Reply:
Done.